File: python_backend/src/services/platforms/tiktok_service.py
"""
TikTok Service
Production-ready TikTok API v2 client using OAuth 2.0
Handles video publishing and authentication
Uses tiktok-api-client library (latest 2025 version)
"""
import httpx
import asyncio
from typing import Optional, Dict, Any
from datetime import datetime

from ...config import settings


class TikTokService:
    """TikTok API service for video publishing"""
    
    # API Constants
    TIKTOK_OAUTH_URL = "https://www.tiktok.com/v2/auth/authorize/"
    TIKTOK_TOKEN_URL = "https://open.tiktokapis.com/v2/oauth/token/"
    TIKTOK_USER_INFO_URL = "https://open.tiktokapis.com/v2/user/info/"
    TIKTOK_VIDEO_PUBLISH_URL = "https://open.tiktokapis.com/v2/post/publish/video/init/"
    TIKTOK_PUBLISH_STATUS_URL = "https://open.tiktokapis.com/v2/post/publish/status/fetch/"
    
    def __init__(self):
        self.http_client = httpx.AsyncClient(timeout=120.0)  # Longer timeout for video uploads
    
    async def close(self):
        """Close HTTP client"""
        await self.http_client.aclose()
    
    # ============================================================================
    # TOKEN MANAGEMENT
    # ============================================================================
    
    async def refresh_access_token(
        self,
        refresh_token: str,
        client_key: str,
        client_secret: str
    ) -> Dict[str, Any]:
        """
        Refresh TikTok access token
        TikTok tokens expire in ~24 hours
        
        Args:
            refresh_token: Refresh token
            client_key: TikTok client key
            client_secret: TikTok client secret
            
        Returns:
            Dict with access_token, refresh_token, expires_in
        """
        try:
            response = await self.http_client.post(
                self.TIKTOK_TOKEN_URL,
                data={
                    'client_key': client_key,
                    'client_secret': client_secret,
                    'refresh_token': refresh_token,
                    'grant_type': 'refresh_token'
                },
                headers={'Content-Type': 'application/x-www-form-urlencoded'}
            )
            
            response.raise_for_status()
            data = response.json()
            
            return {
                'success': True,
                'access_token': data['access_token'],
                'refresh_token': data.get('refresh_token', refresh_token),
                'expires_in': data['expires_in']
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    # ============================================================================
    # USER INFO
    # ============================================================================
    
    async def get_user_info(self, access_token: str) -> Dict[str, Any]:
        """
        Get TikTok user profile information
        
        Args:
            access_token: Access token
            
        Returns:
            Dict with user info
        """
        try:
            response = await self.http_client.get(
                self.TIKTOK_USER_INFO_URL,
                params={'fields': 'open_id,union_id,display_name,avatar_url'},
                headers={'Authorization': f'Bearer {access_token}'}
            )
            
            response.raise_for_status()
            data = response.json()
            
            user_data = data.get('data', {}).get('user', {})
            
            return {
                'success': True,
                'open_id': user_data.get('open_id'),
                'union_id': user_data.get('union_id'),
                'display_name': user_data.get('display_name'),
                'avatar_url': user_data.get('avatar_url')
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    # ============================================================================
    # VIDEO PUBLISHING
    # ============================================================================
    
    async def init_video_publish(
        self,
        access_token: str,
        title: str,
        video_url: str,
        privacy_level: str = "SELF_ONLY"
    ) -> Dict[str, Any]:
        """
        Initialize video publish using PULL_FROM_URL method
        
        TikTok will pull the video from the provided URL
        Important: URL must be from a verified domain
        
        Args:
            access_token: Access token
            title: Video title/caption (max 2200 chars)
            video_url: Publicly accessible video URL
            privacy_level: Privacy level (PUBLIC_TO_EVERYONE, MUTUAL_FOLLOW_FRIENDS, SELF_ONLY)
            
        Returns:
            Dict with publish_id
        """
        try:
            body = {
                'post_info': {
                    'title': title,
                    'privacy_level': privacy_level,
                    'disable_comment': False,
                    'disable_duet': False,
                    'disable_stitch': False
                },
                'source_info': {
                    'source': 'PULL_FROM_URL',
                    'video_url': video_url
                }
            }
            
            response = await self.http_client.post(
                self.TIKTOK_VIDEO_PUBLISH_URL,
                json=body,
                headers={
                    'Authorization': f'Bearer {access_token}',
                    'Content-Type': 'application/json'
                }
            )
            
            response.raise_for_status()
            data = response.json()
            
            # Check for errors
            if data.get('error') and data['error'].get('code') and data['error']['code'] != 'ok':
                error_code = data['error'].get('code')
                error_message = data['error'].get('message')
                log_id = data['error'].get('log_id')
                
                return {
                    'success': False,
                    'error': f"TikTok API error: code={error_code}, message={error_message}, log_id={log_id}"
                }
            
            publish_id = data.get('data', {}).get('publish_id')
            
            if not publish_id:
                return {'success': False, 'error': 'No publish_id returned from TikTok'}
            
            return {
                'success': True,
                'publish_id': publish_id
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    async def init_video_publish_file_upload(
        self,
        access_token: str,
        title: str,
        video_size: int,
        privacy_level: str = "SELF_ONLY"
    ) -> Dict[str, Any]:
        """
        Initialize video publish using FILE_UPLOAD method
        
        Returns upload_url for chunked upload
        
        Args:
            access_token: Access token
            title: Video title/caption
            video_size: Video file size in bytes
            privacy_level: Privacy level
            
        Returns:
            Dict with publish_id and upload_url
        """
        try:
            chunk_size = min(video_size, 10000000)  # Max 10MB chunks
            total_chunks = (video_size + chunk_size - 1) // chunk_size
            
            body = {
                'post_info': {
                    'title': title,
                    'privacy_level': privacy_level,
                    'disable_comment': False,
                    'disable_duet': False,
                    'disable_stitch': False
                },
                'source_info': {
                    'source': 'FILE_UPLOAD',
                    'video_size': video_size,
                    'chunk_size': chunk_size,
                    'total_chunk_count': total_chunks
                }
            }
            
            response = await self.http_client.post(
                self.TIKTOK_VIDEO_PUBLISH_URL,
                json=body,
                headers={
                    'Authorization': f'Bearer {access_token}',
                    'Content-Type': 'application/json'
                }
            )
            
            response.raise_for_status()
            data = response.json()
            
            # Check for errors
            if data.get('error') and data['error'].get('code') and data['error']['code'] != 'ok':
                error_message = data['error'].get('message')
                return {'success': False, 'error': f"TikTok API error: {error_message}"}
            
            publish_id = data.get('data', {}).get('publish_id')
            upload_url = data.get('data', {}).get('upload_url')
            
            if not publish_id or not upload_url:
                return {'success': False, 'error': 'Missing publish_id or upload_url'}
            
            return {
                'success': True,
                'publish_id': publish_id,
                'upload_url': upload_url
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}
    
    async def check_publish_status(
        self,
        access_token: str,
        publish_id: str
    ) -> Dict[str, Any]:
        """
        Check video publish status
        
        Args:
            access_token: Access token
            publish_id: Publish ID from init_video_publish
            
        Returns:
            Dict with status
        """
        try:
            response = await self.http_client.post(
                self.TIKTOK_PUBLISH_STATUS_URL,
                json={'publish_id': publish_id},
                headers={
                    'Authorization': f'Bearer {access_token}',
                    'Content-Type': 'application/json'
                }
            )
            
            response.raise_for_status()
            data = response.json()
            
            status = data.get('data', {}).get('status', 'UNKNOWN')
            
            return {
                'success': True,
                'status': status,
                'publish_id': publish_id
            }
            
        except Exception as e:
            return {'success': False, 'error': str(e)}


# Singleton instance
tiktok_service = TikTokService()


# Helper function
async def close_tiktok_service():
    """Close TikTok service HTTP client"""
    await tiktok_service.close()

File: src/services/platforms/tiktokService.ts
/**
 * TIKTOK SERVICE
 * Implementation of TikTok API v2 integration
 * OAuth 2.0 for content creator access
 * Latest API as of 2025
 */

import { BasePlatformService } from './BasePlatformService'
import {
  OAuthCallbackData,
  OAuthTokenResponse,
  OAuthUserProfile,
  PlatformCredentials,
  PlatformPost,
  PlatformPostResponse,
  PlatformAnalytics,
  PlatformMedia,
  PLATFORM_CONFIGS,
  OAUTH_SCOPES
} from '@/core/types/PlatformTypes'
import type { TikTokCredentials } from '@/types'
import { ExternalAPIError } from '@/core/errors/AppError'

/**
 * TikTok API v2 Implementation
 * Documentation: https://developers.tiktok.com/doc/tiktok-api
 * Note: TikTok API has strict limitations on direct video posting
 * Most operations require TikTok app integration or server-side uploads
 */
export class TikTokService extends BasePlatformService {
  private apiBaseUrl = 'https://open.tiktokapis.com/v2'

  constructor() {
    super('tiktok', PLATFORM_CONFIGS.tiktok.name, PLATFORM_CONFIGS.tiktok.icon)
  }

  /**
   * Generate OAuth authorization URL
   */
  getAuthorizationUrl(state: string, codeChallenge?: string): string {
    const params = new URLSearchParams({
      client_key: this.config.clientId,
      response_type: 'code',
      scope: this.config.scopes.join(','),
      redirect_uri: this.config.redirectUri,
      state
    })

    return `https://www.tiktok.com/v2/auth/authorize/?${params.toString()}`
  }

  /**
   * Exchange authorization code for access token (v2 API)
   */
  async exchangeCodeForToken(callbackData: OAuthCallbackData): Promise<OAuthTokenResponse> {
    try {
      const response = await fetch(`${this.apiBaseUrl}/oauth/token/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Cache-Control': 'no-cache'
        },
        body: new URLSearchParams({
          client_key: this.config.clientId,
          client_secret: this.config.clientSecret,
          code: callbackData.code,
          grant_type: 'authorization_code',
          redirect_uri: this.config.redirectUri
        }).toString()
      })

      if (!response.ok) {
        const error = await response.json()
        throw new ExternalAPIError('TikTok', `Token exchange failed: ${error.error_description || error.error}`)
      }

      const data = await response.json()

      return {
        accessToken: data.access_token,
        refreshToken: data.refresh_token,
        expiresIn: data.expires_in || 7200,
        tokenType: data.token_type || 'Bearer'
      }
    } catch (error) {
      this.handleError(error, 'Token exchange')
    }
  }

  /**
   * Refresh access token using refresh token (v2 API)
   */
  async refreshAccessToken(refreshToken: string): Promise<OAuthTokenResponse> {
    try {
      const response = await fetch(`${this.apiBaseUrl}/oauth/token/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
          'Cache-Control': 'no-cache'
        },
        body: new URLSearchParams({
          client_key: this.config.clientId,
          client_secret: this.config.clientSecret,
          refresh_token: refreshToken,
          grant_type: 'refresh_token'
        }).toString()
      })

      if (!response.ok) {
        const error = await response.json()
        throw new ExternalAPIError('TikTok', `Token refresh failed: ${error.error_description || error.error}`)
      }

      const data = await response.json()

      return {
        accessToken: data.access_token,
        refreshToken: data.refresh_token || refreshToken,
        expiresIn: data.expires_in || 7200,
        tokenType: data.token_type || 'Bearer'
      }
    } catch (error) {
      this.handleError(error, 'Token refresh')
    }
  }

  /**
   * Get authenticated user profile (v2 API)
   */
  async getUserProfile(accessToken: string): Promise<OAuthUserProfile> {
    try {
      const response = await fetch(`${this.apiBaseUrl}/user/info/?fields=open_id,union_id,display_name,avatar_url`, {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${accessToken}`
        }
      })

      if (!response.ok) {
        throw new ExternalAPIError('TikTok', `Failed to fetch user profile: ${response.status}`)
      }

      const data = await response.json()
      const user = data.data.user

      return {
        id: user.open_id,
        username: user.display_name,
        name: user.display_name
      }
    } catch (error) {
      this.handleError(error, 'Get user profile')
    }
  }

  /**
   * Post content to TikTok
   * Note: Direct posting via API requires business account and complex upload process
   */
  async postContent(
    credentials: PlatformCredentials,
    post: PlatformPost
  ): Promise<PlatformPostResponse> {
    try {
      if (post.content.length > PLATFORM_CONFIGS.tiktok.maxCharacters) {
        return this.formatErrorResponse(
          new Error(`Content exceeds ${PLATFORM_CONFIGS.tiktok.maxCharacters} characters`),
          'Post content'
        )
      }

      // TikTok requires video for posting
      if (!post.media || post.media.length === 0) {
        return this.formatErrorResponse(
          new Error('TikTok requires video content to post'),
          'Post content'
        )
      }

      return this.formatErrorResponse(
        new Error('Direct TikTok posting requires additional setup. Use TikTok Creator Studio instead.'),
        'Post content'
      )
    } catch (error) {
      return this.formatErrorResponse(error, 'Post content')
    }
  }

  /**
   * Upload media to TikTok
   */
  async uploadMedia(
    credentials: PlatformCredentials,
    media: PlatformMedia
  ): Promise<string> {
    try {
      // Fetch media from URL
      const mediaResponse = await fetch(media.url)
      if (!mediaResponse.ok) {
        throw new Error('Failed to fetch media from URL')
      }

      const mediaBuffer = await mediaResponse.arrayBuffer()

      // Validate media size (287MB max for TikTok)
      if (mediaBuffer.byteLength > PLATFORM_CONFIGS.tiktok.maxMediaSize) {
        throw new Error(`Media exceeds ${PLATFORM_CONFIGS.tiktok.maxMediaSize} bytes`)
      }

      return media.url
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error)
      throw new ExternalAPIError('TikTok', `Media upload failed: ${message}`)
    }
  }

  /**
   * Schedule post - TikTok doesn't support scheduled posting via API
   */
  async schedulePost(
    credentials: PlatformCredentials,
    post: PlatformPost,
    scheduledTime: Date
  ): Promise<PlatformPostResponse> {
    return {
      postId: '',
      platform: 'tiktok',
      status: 'failed',
      error: 'TikTok does not support scheduled posting via API. Use TikTok Creator Studio.',
      createdAt: new Date()
    }
  }

  /**
   * Verify credentials (v2 API)
   */
  async verifyCredentials(credentials: PlatformCredentials): Promise<boolean> {
    try {
      const response = await fetch(`${this.apiBaseUrl}/user/info/?fields=open_id`, {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${credentials.accessToken}`
        }
      })

      return response.ok && response.status === 200
    } catch (error) {
      return false
    }
  }

  /**
   * Get post metrics from TikTok Analytics
   */
  async getPostMetrics(
    credentials: PlatformCredentials,
    postId: string
  ): Promise<PlatformAnalytics> {
    try {
      const response = await fetch(`${this.apiBaseUrl}/video/query/`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${credentials.accessToken}`,
          'User-Agent': 'SocialMediaOS/1.0'
        },
        body: JSON.stringify({
          filters: {
            video_ids: [postId]
          },
          fields: ['like_count', 'comment_count', 'share_count', 'view_count']
        })
      })

      if (!response.ok) {
        throw new Error('Failed to fetch video metrics')
      }

      const data = await response.json()
      const video = data.data.videos[0]

      return {
        postId,
        platform: 'tiktok',
        impressions: video.view_count || 0,
        views: video.view_count || 0,
        likes: video.like_count || 0,
        comments: video.comment_count || 0,
        shares: video.share_count || 0,
        engagements:
          (video.like_count || 0) + (video.comment_count || 0) + (video.share_count || 0),
        fetched_at: new Date()
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error)
      throw new ExternalAPIError('TikTok', `Failed to fetch metrics: ${message}`)
    }
  }

  /**
   * Get platform max character limit
   */
  getMaxCharacterLimit(): number {
    return PLATFORM_CONFIGS.tiktok.maxCharacters
  }

  /**
   * Check if platform supports scheduling
   */
  supportsScheduling(): boolean {
    return PLATFORM_CONFIGS.tiktok.supportsScheduling
  }

  /**
   * Check if platform supports media upload
   */
  supportsMediaUpload(): boolean {
    return PLATFORM_CONFIGS.tiktok.supportsMediaUpload
  }
}

/**
 * Upload video to storage (returns public URL)
 * Backend endpoint: /api/tiktok/upload-media
 * Returns public URL for use in TikTok API
 */
export async function uploadTikTokVideo(
  credentials: TikTokCredentials,
  videoData: string  // base64 or public URL
): Promise<{ success: boolean; videoUrl?: string; videoSize?: number; error?: string }> {
  try {
    // If it's already a public URL, return it directly
    // TikTok API module doesn't have uploadMedia, so we just pass through URLs
    if (videoData.startsWith('http')) {
      return {
        success: true,
        videoUrl: videoData,
        videoSize: 0
      };
    }

    // For base64 data, we need to upload to storage first
    // This would typically be handled by a storage upload API
    return {
      success: false,
      error: 'Base64 upload not supported. Please provide a public URL.'
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Upload failed'
    };
  }
}

/**
 * Get TikTok account info via backend API
 */
export async function getTikTokAccountInfo(
  credentials: TikTokCredentials
): Promise<{ success: boolean; accountInfo?: any; error?: string }> {
  try {
    if (!credentials.isConnected) {
      return { success: false, error: 'TikTok account not connected' };
    }

    // Use Python backend client for TikTok verification
    const { verifyCredentials } = await import('@/lib/python-backend/api/social/tiktok');

    const result = await verifyCredentials();

    return {
      success: result.success,
      accountInfo: {
        accountId: result.accountId,
        accountName: result.accountName,
        connected: result.connected,
        expiresAt: result.expiresAt
      }
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Failed to fetch account info'
    };
  }
}

/**
 * Post to TikTok via backend API
 * TikTok Content Posting API
 * Supports video posts with caption (2200 char limit)
 */
export async function postToTikTok(
  credentials: any,
  options: { caption: string; videoUrl: string; videoSize: number }
): Promise<{ success: boolean; videoId?: string; url?: string; error?: string }> {
  try {
    // Use Python backend client for TikTok posting
    const { createPost } = await import('@/lib/python-backend/api/social/tiktok');

    const result = await createPost({
      caption: options.caption,
      videoUrl: options.videoUrl,
    });

    return {
      success: result.success,
      videoId: result.publishId,
      url: undefined // TikTok API doesn't return a direct URL
    };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Failed to post to TikTok'
    };
  }
}
